File: csharp/HearthstoneReplays/Parser/Handlers/OptionsHandler.cs
#region

using System;
using System.Collections.Generic;
using HearthstoneReplays.Hearthstone.Enums;
using HearthstoneReplays.ReplayData;
using HearthstoneReplays.ReplayData.Meta.Options;

#endregion

namespace HearthstoneReplays.Parser.Handlers
{
	public class OptionsHandler
	{
		public static void Handle(string timestamp, string data, ParserState state)
		{
			data = data.Trim();
			var match = Regexes.OptionsEntityRegex.Match(data);
			if(match.Success)
			{
				var id = match.Groups[1].Value;
				state.Options = new Options {Id = int.Parse(id), OptionList = new List<Option>()};
				if(state.Node.Type == typeof(Game))
					((Game)state.Node.Object).Data.Add(state.Options);
				else
					throw new Exception("Invalid node " + state.Node.Type + " -- " + data);
				return;
			}
			match = Regexes.OptionsOptionRegex.Match(data);
			if(match.Success)
			{
				var index = match.Groups[1].Value;
				var rawType = match.Groups[2].Value;
				var rawEntity = match.Groups[3].Value;
				var entity = Helper.ParseEntity(rawEntity, state);
			    var type = Helper.ParseEnum<OPTION_TYPE>(rawType);
				var option = new Option {Entity = entity, Index = int.Parse(index), Type = type, OptionItems = new List<OptionItem>()};
				state.Options.OptionList.Add(option);
				state.CurrentOption = option;
				state.LastOption = option;
				return;
			}
			match = Regexes.OptionsSuboptionRegex.Match(data);
			if(match.Success)
			{
				var subOptionType = match.Groups[1].Value;
				var index = match.Groups[2].Value;
				var rawEntity = match.Groups[3].Value;
				var entity = Helper.ParseEntity(rawEntity, state);

				if(subOptionType == "subOption")
				{
					var subOption = new SubOption {Entity = entity, Index = int.Parse(index), Targets = new List<Target>()};
					state.CurrentOption.OptionItems.Add(subOption);
					state.LastOption = subOption;
				}
				else if(subOptionType == "target")
				{
					var target = new Target {Entity = entity, Index = int.Parse(index)};
					var lastOption = state.LastOption as Option;
					if(lastOption != null)
					{
						lastOption.OptionItems.Add(target);
						return;
					}
					var lastSubOption = state.LastOption as SubOption;
					if(lastSubOption != null)
						lastSubOption.Targets.Add(target);
				}
				else
					throw new Exception("Unexpected suboption type: " + subOptionType);
			}
		}
	}
}

File: java/src/main/java/info/hearthsim/hsreplay/parser/handlers/OptionsHandler.java
package info.hearthsim.hsreplay.parser.handlers;

import info.hearthsim.hsreplay.enums.OptionType;
import info.hearthsim.hsreplay.parser.Helper;
import info.hearthsim.hsreplay.parser.ParserState;
import info.hearthsim.hsreplay.parser.Regexes;
import info.hearthsim.hsreplay.parser.replaydata.Game;
import info.hearthsim.hsreplay.parser.replaydata.meta.options.Option;
import info.hearthsim.hsreplay.parser.replaydata.meta.options.Options;
import info.hearthsim.hsreplay.parser.replaydata.meta.options.SubOption;
import info.hearthsim.hsreplay.parser.replaydata.meta.options.Target;

import java.util.ArrayList;
import java.util.regex.Matcher;

public class OptionsHandler {

	public static void handle(String timestamp, String data, ParserState state) throws Exception {
		data = data.trim();
		Matcher match = Regexes.OptionsEntityRegex.matcher(data);
		if (match.matches()) {
			String id = match.group(1);
			Options options = new Options(timestamp, Integer.parseInt(id), new ArrayList<>());
			state.setOptions(options);
			state.updateCurrentNode(Game.class);

			if (state.getNode().getType().isAssignableFrom(Game.class))
				((Game) state.getNode().getObject()).getData().add(options);
			else
				throw new Exception("Invalid node " + state.getNode().getType());

			return;
		}

		match = Regexes.OptionsOptionRegex.matcher(data);
		if (match.matches()) {
			String index = match.group(1);
			String rawType = match.group(2);
			String rawEntity = match.group(3);

			int entity = Helper.parseEntity(rawEntity, state);
			int type = OptionType.parseEnum(rawType);

			Option option = new Option(Integer.parseInt(index), type, entity, new ArrayList<>());
			state.getOptions().getOptionList().add(option);
			state.setCurrentOption(option);
			state.setLastOption(option);
			return;
		}

		match = Regexes.OptionsSuboptionRegex.matcher(data);
		if (match.matches()) {
			String subOptionType = match.group(1);
			String index = match.group(2);
			String rawEntity = match.group(3);
			int entity = Helper.parseEntity(rawEntity, state);

			if ("subOption".equals(subOptionType)) {
				SubOption subOption = new SubOption(entity, Integer.parseInt(index), new ArrayList<>());
				state.getCurrentOption().getOptionsItems().add(subOption);
				state.setLastOption(subOption);
			}
			else if ("target".equals(subOptionType)) {
				Target target = new Target(entity, Integer.parseInt(index));
				Option lastOption = state.getLastOption() instanceof Option ? (Option) state.getLastOption() : null;
				if (lastOption != null) {
					lastOption.getOptionsItems().add(target);
					return;
				}
				SubOption lastSubOption = state.getLastOption() instanceof SubOption ? (SubOption) state
						.getLastOption() : null;
				if (lastSubOption != null) {
					lastSubOption.getTargets().add(target);
				}
			}
			else
				throw new Exception("Unexpected suboption type: " + subOptionType);
		}

	}

}
